### ISSUE TITLE
rpc: "missing address" 

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#29035]
Forked from https://github.com/cockroachdb/cockroach/issues/27731#issuecomment-410734188:

During the gossip partition, the logs showed a very large number of errors of the form: `message":"grpc: addrConn.createTransport failed to connect to { 0 <nil>}. Err :connection error: desc = \"transport: Error while dialing dial tcp: missing address\". Reconnecting...`

Looking though the grpc code, this appears to be caused by passing an empty string to `RPCContext.GRPCDial()`. These errors went away when the gossip partition was resolved. Perhaps related, this occurred in a 9 node cluster which had 3 decommission nodes (i.e. there were 9 live nodes and 3 nodes had previously been decommissioned and removed). 

Current theory is that something is gossiping an empty node descriptor or a node descriptor with an empty address field, but so far code inspection hasn't revealed how that could occur and attempts at reproduction of been fruitless.

**Comment by petermattis** [comment]
I've been spelunking the code for the last hour and don't have any theory as to how an empty node descriptor could be gossiped. Well, that's not quite true. There is a known way for an empty node descriptor to be gossiped, if a node is decommissioned and wiped and then a new node is started with the same advertise address. What I'm failing to see is why anything would be trying to open a connection to the old (decommissioned) node ID. And why would the timing of such connection attempts correspond to a gossip partition.

**Comment by petermattis** [comment]
I'm not sure this is an S-anything bug. There isn't a user-visible impact here (as far as I know).

**Comment by petermattis** [comment]
(I'd still like to track this down, but it isn't necessary for 2.1).

**Comment by tbg** [comment]
Isn't this just

https://github.com/cockroachdb/cockroach/blob/18688614253b76c230eedb2a6e9aad80785b4522/pkg/gossip/gossip.go#L848-L860

I've been annoyed by that code before. The empty proto leaks into all consumers of Gossip. Feels like we should be able to teach Gossip that you can gossip an empty bytes slice as sort of a "tombstone" instead.

**Comment by petermattis** [comment]
Oh, good point. I always forget about this code.

**Comment by petermattis** [comment]
Note that we already have some detection of this empty node descriptors:

```
		// Don't return node descriptors that are empty, because that's meant to
		// indicate that the node has been removed from the cluster.
		if nodeDescriptor.NodeID == 0 || nodeDescriptor.Address.IsEmpty() {
			return nil, errors.Errorf("n%d has been removed from the cluster", nodeID)
		}
```

**Comment by petermattis** [comment]
Hmm, it's likely problematic that the empty description detection code is preceded by:

```
func (g *Gossip) getNodeDescriptorLocked(nodeID roachpb.NodeID) (*roachpb.NodeDescriptor, error) {
	if desc, ok := g.nodeDescs[nodeID]; ok {
		return desc, nil
	}
  ...
```

So in most (all) cases it is having no effect.

**Comment by petermattis** [comment]
Hmm, my previous comment was correct, but similar code exists in `Gossip.updateNodeAddress` which is the only place that `g.nodeDescs` is set:

```
	if desc.NodeID == 0 && desc.Address.IsEmpty() {
		nodeID, err := NodeIDFromKey(key, KeyNodeIDPrefix)
		if err != nil {
			log.Errorf(ctx, "unable to update node address for removed node: %s", err)
			return
		}
		log.Infof(ctx, "removed n%d from gossip", nodeID)
		g.removeNodeDescriptorLocked(nodeID)
		return
	}

	existingDesc, ok := g.nodeDescs[desc.NodeID]
	if !ok || !proto.Equal(existingDesc, &desc) {
		g.nodeDescs[desc.NodeID] = &desc
	}
```

Note that there is a slight discrepancy between these code paths. One uses `if desc.NodeID == 0 && desc.Address.IsEmpty()` while the other uses `if nodeDescriptor.NodeID == 0 || nodeDescriptor.Address.IsEmpty()`. I'm not seeing how that would cause a problem, though.

I'm also still not seeing any way for an empty address to be used.

**Comment by petermattis** [comment]
This was fixed by #33171, though at the time it wasn't clear because we didn't know how to reproduce that message. #34120 shows how to reproduce.

**Comment by shuwenhe** [comment]
what's the matter ?
rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp: missing address"
